sql/2021/pwa/lighthouse_pwa_audits.sql

```diff
@@ -0,0 +1,42 @@
+#standardSQL
+# Get summary of all lighthouse scores for a category
+# Note scores, weightings, groups and descriptions may be off in mixed months when new versions of Lighthouse roles out
+
+CREATE TEMPORARY FUNCTION getAudits(report STRING, category STRING)
+RETURNS ARRAY<STRUCT<id STRING, weight INT64, audit_group STRING, title STRING, description STRING, score INT64>> LANGUAGE js AS '''
+var $ = JSON.parse(report);
+var auditrefs = $.categories[category].auditRefs;
+var audits = $.audits;
+$ = null;
+var results = [];
+for (auditref of auditrefs) {
+  results.push({
+    id: auditref.id,
+    weight: auditref.weight,
+    audit_group: auditref.group,
+    description: audits[auditref.id].description,
+    score: audits[auditref.id].score
+  });
+}
+return results;
+''';
+
+SELECT
+  audits.id AS id,
+  COUNTIF(audits.score > 0) AS num_pages,
+  COUNT(0) AS total,
+  COUNTIF(audits.score IS NOT NULL) AS total_applicable,
+  SAFE_DIVIDE(COUNTIF(audits.score > 0), COUNTIF(audits.score IS NOT NULL)) AS pct,
+  APPROX_QUANTILES(audits.weight, 100)[OFFSET(50)] AS median_weight,
+  MAX(audits.audit_group) AS audit_group,
+  MAX(audits.description) AS description
+FROM
+  `httparchive.lighthouse.2021_07_01_mobile`,
+  UNNEST(getAudits(report, "pwa")) AS audits
+WHERE
+  LENGTH(report) < 20000000  # necessary to avoid out of memory issues. Excludes 16 very large results
+GROUP BY
+  audits.id
+ORDER BY
+  median_weight DESC,
+  id
```

Comment:
Compared with 2020, Lighthouse has removed the **load-fast-enough-for-pwa** and **works-offline** ([issue](https://github.com/GoogleChrome/lighthouse/pull/11806)). Those used to be available in last year’s `sample_data_2020` dataset.

For this year I wanted to check the YoY growth of sites that respond offline, but that won't be possible.
In any case, this query LGTM.

Reply:
Actually that reminds me I want to rewrite this a bit [after learnings elsewhere](https://github.com/HTTPArchive/bigquery/pull/114).

Currently we take the whole Lighthouse JSON and sent it to the JavaScript UDF for processing. This causes that to fail for large payloads - hence the WHERE clause to exclude this.

But it’s more efficient to JSON_EXTRACT the bit we want (PWA audits) and then send that smaller JSON to the UDF instead. Then don’t need that WHERE clause. Plus it runs a lot faster (in that other query at least so guessing same here). I guess JSON_EXTRACT is more efficient than passing large objects between SQL and JavaScript.

Will leave this comment open as a reminder to myself to refactor this a little before merging.

Reply:
Rewrote these to be more efficient in https://github.com/HTTPArchive/almanac.httparchive.org/pull/2272/commits/9c83af78c0987c6a6a7e105c30faa54ae283ef5c. Resolving.